File: leetcode/Binary_Tree/0637_Average_of_Levels_in_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    vector<long long> sum;
    vector<int> count;
    void calculateNode(int level, TreeNode *node) {
        if (node == NULL)
            return;
        int n = sum.size();
        if (level > n - 1) {
            sum.push_back(node -> val);
            count.push_back(1);
        }
        else {
            sum[level] += (node -> val);
            count[level] ++;
        }
        calculateNode(level + 1, node -> left);
        calculateNode(level + 1, node -> right);
    }
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> result;
        calculateNode(0, root);
        for (int i = 0; i < sum.size(); i++) {
            result.push_back((double)sum[i] / (double)count[i]);
        }
        return result;
    }
};

File: leetcode/Binary_Tree/0637_Average_of_Levels_in_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def averageOfLevels(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[float]
        """

        result = []
        current_level = [root]
        while current_level:
            node_sum = 0
            next_level = []
            for node in current_level:
                node_sum += node.val
                if node.left:
                    next_level.append(node.left)
                if node.right:
                    next_level.append(node.right)

            result.append(float(node_sum) / len(current_level))
            current_level = next_level

        return result
